src/bp/core/services/migration/index.ts

```diff
@@ -0,0 +1,197 @@
+        return this.logger.info(`Skipping already migrated file "${file}"`)
+      }
+
+      this.logger.info(`Running ${file}`)
+
+      const result = await this.loadedMigrations[file].up(api, this.configProvider, this.database, container)
+      if (result.success) {
+        this._saveCompletedMigration(file, result)
+        await this.logger.info(`- ${result.message || 'Success'}`)
+      } else {
+        hasFailures = true
+        await this.logger.error(`- ${result.message || 'Failure'}`)
+      }
+    })
+
+    if (hasFailures) {
+      await this.logger.error(
+        `Some steps failed to complete. Please fix errors manually, then restart Botpress so the update process may finish.`
+      )
+      process.exit(1)
+    }
+
+    await this.configProvider.mergeBotpressConfig({ version: this.currentVersion })
+    this.logger.info(`Migrations completed successfully! `)
+  }
+
+  private _getCompletedMigrations(): string[] {
+    return fs.readdirSync(this.completedMigrationsDir)
+  }
+
+  private _saveCompletedMigration(filename: string, result: sdk.MigrationResult) {
+    mkdirp.sync(this.completedMigrationsDir)
+    fs.writeFileSync(path.resolve(`${this.completedMigrationsDir}/${filename}`), JSON.stringify(result, undefined, 2))
+  }
+
+  private _loadMigrations = (fileList: MigrationFile[]) =>
+    fileList.map(file => (this.loadedMigrations[file.filename] = require(file.location).default))
+
+  private _getMissingCoreMigrations() {
+    return this._getFilteredMigrations(path.join(__dirname, '../../../migrations'))
+  }
+
+  private _getMissingModuleMigrations() {
+    return _.flatMap(Object.keys(process.LOADED_MODULES), module =>
+      this._getFilteredMigrations(path.join(process.LOADED_MODULES[module], 'dist/migrations'))
+    )
+  }
+
+  private _getFilteredMigrations = (rootPath: string) => this._filterMissing(this._getMigrations(rootPath))
+
+  private _getMigrations(rootPath: string): MigrationFile[] {
+    return _.orderBy(
+      glob.sync('**/*.js', { cwd: rootPath }).map(filepath => {
+        const [rawVersion, timestamp, title] = path.basename(filepath).split('-')
+        return {
+          filename: path.basename(filepath),
+          version: semver.valid(rawVersion.replace(/_/g, '.')),
+          title: (title || '').replace('.js', ''),
+          date: Number(timestamp),
+          location: path.join(rootPath, filepath)
+        }
+      }),
+      'date'
+    )
+  }
+
+  private _filterMissing = (files: MigrationFile[]) =>
+    files.filter(file => semver.satisfies(file.version, `>${this.configVersion} <= ${this.currentVersion}`))
+}
+
+interface MigrationFile {
+  date: number
+  version: string
+  location: string
+  filename: string
+  title: string
+}
+
+export interface Migration {
+  info: {
+    description: string
+    type: 'database' | 'config' | 'content'
+  }
+  up: (bp: typeof sdk, config: ConfigProvider, database: Database, inversify: Container) => Promise<sdk.MigrationResult>
+  down?: (
```

Comment:
I think I'd make `down` not optional. Let's say we implement the rollback functionality and we can't rollback because some migration scripts don't have a `down`. We'd have to manually apply the rollback.

Reply:
Let's say a migration is breaking for some reason, the client wants to rollback to its previous state but can't

Reply:
I was thinking of just removing down for now. Theres no logic in place to handle it, it's just to keep the signature for future ones